### ISSUE TITLE
Add to cart does not send product_name to the API

### ISSUE THREAD

**Issue by joshuaiz** [co-cart/co-cart#34]
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] The issue still exists against the latest `master` branch of Cart REST-API for WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue.
- [x] I have included a failing test as a pull request (Optional)
- [x] I have installed the requirements to run this plugin.

## Steps to reproduce the issue

1. Make Add to Cart request using the `cart/add` endpoint
2. console.log(response)
3. `product_name` is not returned in the JSON data

I thought my request was configured improperly at first but I looked at the `class-wc-rest-cart-controller.php` file and in the `add_to_cart` function starting on line 276, `product_name` is never added to the `$data` array, so it is never sent and thus not returned in the original response. 

It looks like WooCommerce figures out which product to add to the cart solely by the id so when you reload the page and get the current cart contents, `product_name` is returned. But, I'm trying to update my cart with the initial response and the name is missing :)

Alternatively, I could send this with the `data` parameter but it is not clear how that needs to be formatted in the docs. I tried sending an object, an array, or individual value and could not get it to work. In the docs it says the parameter is `data` but in the `class-wc-rest-cart-controller.php` file it looks like the parameter is `cart_item_data`. How do we add additional data to the Add To Cart request?

Thanks!

## Expected/actual behaviour

When I follow those steps, I see...

no `product_name` value in the initial response from the Add To Cart endpoint

I was expecting to see...

`product_name` there.

## Isolating the problem

- [x] This bug happens with only WooCommerce and Cart REST-API for WooCommerce plugin are active.
- [x] This bug happens with a default WordPress theme active.
- [x] This bug happens with the WordPress theme Storefront active.
- [x] This bug happens with the latest release of WooCommerce active.
- [x] I can reproduce this bug consistently using the steps above.

## WordPress Environment

Not WordPress related per se...I'm making the requests client-side.

<details>
```
Go to "WooCommerce > System Status then copy and paste the system status report here.
```
</details>

**Comment by seb86** [comment]
Hi @joshuaiz 

> It looks like WooCommerce figures out which product to add to the cart solely by the id so when you reload the page and get the current cart contents, product_name is returned. But, I'm trying to update my cart with the initial response and the name is missing :)

The add to cart function is exactly the same as to how WooCommerce does it. The same data, the same message responses, the same filter.

There is an action hook you can use `do_action( 'wc_cart_rest_add_to_cart', $item_key, $data );` to return the product name in the initial response or any data about the product that is not yet returned.

I don't think the product name is actually designed to return in the initial response when adding a product. It might be an error on my part with the documentation. If I am right then I will see about adding the product name in the initial response in the future.

The initial response is returned from the CART class in the core of WC. The only manipulation I have done is when getting the cart contents.

**Comment by joshuaiz** [comment]
Hi @seb86,

Got it. I was banging my head against a wall wondering why the `product_name` wasn't being returned like the docs!

I tried adding data via the hook but I don't think it gets fired when adding something to the cart via the endpoint as nothing triggers it in a headless app. Or am I missing something?

Is there a way to add data to the cart via the `data` parameter using the Add To Cart endpoint? If so, how is that formatted?

In my case I have an 'Add To Cart' button which has an onClick action to trigger an `addToCart` function. From that I am sending the `product_id` and `quantity` like this:

```javascript
<button
    type="button"
    onClick={() => {
        addToCart({
            product_id: id,
            quantity: 1,
            index: index,
            data: {
                name: props.name
            }
        })
    }}
>
    Add To Cart
</button>
```

And then in my `addToCart()` function which accepts the passed data as arguments, my endpoint is like so:

`https://viz.local/wp-json/wc/v2/cart/add?${product_id}&quantity=${quantity}&cart_item_data=${cart_item_data}`

If I include the `cart_item_data` parameter, it doesn't like it — I get a CORS error. If the `cart_item_data` is removed, I can add products to the cart without issue, but of course there's no `product_name`.

If I disable CORS in the browser via an extension, I get this error:

`Warning: is_array() expects exactly 1 parameter, 3 given in /app/public/wp-includes/rest-api/class-wp-rest-request.php on line 858.`

I've tried formatting the `cart_item_data` as an object, as an array or as a string but maybe it needs to be serialized or formatted differently? What am I doing wrong here?

Thanks!

**Comment by seb86** [comment]
> I tried adding data via the hook but I don't think it gets fired when adding something to the cart via the endpoint as nothing triggers it in a headless app. Or am I missing something?

The action hook is not fired via a URL. You have to do that part in PHP.

> Is there a way to add data to the cart via the cart_item_data parameter using the Add To Cart endpoint? If so, how is that formatted?

You may need to serialize your `cart_item_data`.

I don't know what language you are writing in but I can not provide support for anything else but the REST API its self.

**Comment by joshuaiz** [comment]
@seb86 I realize that hook is in PHP - I added it to the functions.php of my theme but It doesn't seem like it was firing when the REST API request was made.

Understood about support.

This is a headless WordPress app with a normal WordPress back end and React front end using the REST API. The snippet above was JSX.

I'll give it a go with trying to serialize the `cart_item_data` and see if that does the trick.

Thanks!

**Comment by joshuaiz** [comment]
Ok I've made a little progress. 

I was trying to serialize on the client end and WordPress didn't like anything I was sending, even after using `JSON.stringify()`. No matter what I did, I kept getting this error:

`Warning: is_array() expects exactly 1 parameter, 3 given in /app/public/wp-includes/rest-api/class-wp-rest-request.php on line 858`

So, I looked in `class-wp-rest-request.php` and could not find anything there that might help me. Just to try, I commented out the function it referred to (starts on line 846):

```php
/*
* Check the validation callbacks for each registered arg.
*
* This is done after required checking as required checking is cheaper.
*/
$invalid_params = array();

foreach ( $attributes['args'] as $key => $arg ) {

    $param = $this->get_param( $key );

    if ( null !== $param && ! empty( $arg['validate_callback'] ) ) {
        $valid_check = call_user_func( $arg['validate_callback'], $param, $this, $key );

        if ( false === $valid_check ) {
            $invalid_params[ $key ] = __( 'Invalid parameter.' );
        }

        if ( is_wp_error( $valid_check ) ) {
            $invalid_params[ $key ] = $valid_check->get_error_message();
        }
    }
}

if ( $invalid_params ) {
    return new WP_Error( 'rest_invalid_param', sprintf( __( 'Invalid parameter(s): %s' ), implode( ', ', array_keys( $invalid_params ) ) ), array( 'status' => 400, 'params' => $invalid_params ) );
}

return true;
```

With that commented out, I can send the data through via the `cart_item_data` parameter and it comes back in the initial response(!):

`addToCart Response {0: "Test Product 2", key: "bfcae80392715c73b997f266b6cca0ce", product_id: 42, variation_id: 0, variation: Array(0), quantity: 1, …}`

^ The `0:` item I'm sending is the product name. Sweetness...kind of.

Obviously I don't want to be editing core files and removing any request validation but I'm at a loss as to what WP/WooCommerce is not liking about my `cart_item_data` parameter. Even if I send a hardcoded string like `&cart_item_data=mystring`, it bugs out. Any ideas on what it doesn't like and how to get it in a format that it does? Does it want an array? An array of objects? A json stringified array/object? 

I guess at this point, I would have a feature request to send the product name with the 'Add To Cart' endpoint which would be ideal :)

**Comment by aMedh** [comment]
$cart_item_data = [
        ["name" => 'Test']
    ];

    $arr = [];

    $data = [
        "product_id" => 1257,
        "quantity" => 1,
        "variation_id" => 1264,
        "variation" => $arr,
        "cart_item_data" => $cart_item_data
      ];
      $results = $woocommerce->post('cart/add', $data);

THis dies not seem to work for. It works if I leave varation and cart_item_data out. Please help.

**Comment by seb86** [comment]
@aMedh You need to pass the data as an array not object.

**Comment by seb86** [comment]
So the add to cart function is pretty much the same as [this function](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-cart.php#L1003). It looks like I need to update my add to cart function to support changes made to the core of WC cart so these variables are passed properly again.

I will create a new branch to push a change and I will let you know when it is ready to test for feedback. 

> No eta as I am on holiday.

**Comment by seb86** [comment]
@joshuaiz Is this issue resolved with the release of v1.0.5?

**Comment by seb86** [comment]
I am going to assume this issue is resolved and close it.
